### PR TITLE
Initial plugin structure

### PR DESCRIPTION
**Pull request by goldenapples**
Adds plugin skeleton: shortcode definition and list of fields to collect in UI.

### CODE REVIEW

**Comment by goldenapples** [comment]
@danielbachhuber Got the plugin structure roughed out; most attributes aren't functional yet, but I wanted to validate the rough structure with you early. Care to review/merge?

File: image-shortcake.php [hunk 1]
```diff
@@ -2,11 +2,73 @@
...
+Description: Provides a shortcode for image elements. Use with the Shortcake plugin for a preview of images
+Author: goldenapples
+Author URI: https://github.com/fusioneng
+Plugin URI: https://github.com/fusioneng/image-shortcake
 Text Domain: image-shortcake
 Domain Path: /languages
 */
 
+require_once( 'inc/class-img-shortcode.php' );
+require_once( 'inc/class-img-shortcode-ui.php' );
```
Comment (danielbachhuber) [inline image-shortcake.php]: Can we put these requires inside a `require_files()` method inside our singleton? Also, you should require with absolute paths, not relative.

File: image-shortcake.php [hunk 2]
```diff
@@ -2,11 +2,73 @@
...
+		}
+		return self::$instance;
+	}
+
+	private function register_shortcodes() {
+
+		if ( ! function_exists( 'shortcode_ui_register_for_shortcode' ) ) {
+			add_action( 'admin_notices', function(){
+				if ( current_user_can( 'activate_plugins' ) ) {
+					echo '<div class="error message"><p>Shortcode UI plugin must be active for Image Shortcake plugin to function.</p></div>';
```
Comment (danielbachhuber) [inline image-shortcake.php]: String should be translatable.

File: image-shortcake.php [hunk 3]
```diff
@@ -2,11 +2,73 @@
...
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new Image_Shortcake;
+			self::$instance->register_shortcodes();
+		}
+		return self::$instance;
+	}
+
+	private function register_shortcodes() {
+
+		if ( ! function_exists( 'shortcode_ui_register_for_shortcode' ) ) {
```
Comment (danielbachhuber) [inline image-shortcake.php]: Given the conditional is relatively small, could we change it to:

```
if ( function_exists ) {
   // register the shortcode
} else {
   // Throw notice
}
```

Also, the shortcode can likely be used independently of the shortcode UI.

File: image-shortcake.php [hunk 4]
```diff
@@ -2,11 +2,73 @@
...
+			add_action( 'admin_notices', function(){
```
Comment (danielbachhuber) [inline image-shortcake.php]: It would be nice to have this as a dedicated method instead of a closure.

File: image-shortcake.php [hunk 5]
```diff
@@ -2,11 +2,73 @@
...
+
+	/**
+	 * Activate the plugin as a singleton.
+	 *
+	 */
+	public static function get_instance() {
+
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new Image_Shortcake;
+			self::$instance->register_shortcodes();
```
Comment (danielbachhuber) [inline image-shortcake.php]: Given we only have one shortcode, we can probably make this singular.

File: inc/class-img-shortcode.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
+<?php
```
Comment (danielbachhuber) [inline inc/class-img-shortcode.php]: File name should follow the class.

**Comment by danielbachhuber** [comment]
> but I wanted to validate the rough structure with you early.

Can we use our existing `Shortcode` class pattern, where UI arguments and render callback are encapsulated into one class?

Also, it would be nice if the tests were passing.

**Comment by goldenapples** [comment]
> Can we use our existing Shortcode class pattern, where UI arguments and render callback are encapsulated into one class?

You mean the [Shortcode class](https://github.com/fusioneng/fusion-theme/blob/master/inc/shortcodes/class-shortcode.php) from the Fusion theme? I was kinda following the pattern in the shortcode-ui plugin thinking that was the pattern to be followed.

Also, a few of the style changes you mention here are things I copied pretty much verbatim from the [sample code in Shortcake](https://github.com/fusioneng/Shortcake/blob/master/dev.php). Would it be worth putting this together for a PR there as well?

**Comment by goldenapples** [comment]
Addressed all the items you mentioned here with the exception of combining the UI arguments and the render function into one class. I'd rather try to keep the static functions (UI arguments) separate from the object-oriented methods (building a shortcode object and rendering it). Does that make sense? If not, it'd be easy enough to merge the two classes at any point.

**Comment by danielbachhuber** [comment]
> I'd rather try to keep the static functions (UI arguments) separate from the object-oriented methods (building a shortcode object and rendering it). Does that make sense?

Well, we already have an existing pattern we can follow. I don't think it makes sense to deviate from the pattern. Let's discuss when you're available though.

**Comment by goldenapples** [comment]
OK, basic structure is in place. @danielbachhuber want to review/merge so I can start new branches for features?

File: image-shortcake.php [hunk 7]
```diff
@@ -2,11 +2,107 @@
...
+
+
+	/**
+	 * Register the [img] shortcode and the UI for it..
+	 *
+	 */
+	private function register_shortcode() {
+
+		add_shortcode( 'img',
+			array( $this, 'shortcake_img_shortcode' )
```
Comment (danielbachhuber) [inline image-shortcake.php]: We can probably drop this method entirely.

File: inc/class-img-shortcode.php [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+	 */
+	public static function get_shortcode_ui_args() {
+
+		$shortcode_ui_args = array(
+
+				'label' => 'Image',
+
+				'listItemImage' => 'dashicons-format-image',
+
+				'post_type'     => array( 'post' ),
```
Comment (danielbachhuber) [inline inc/class-img-shortcode.php]: Shouldn't this be all post types?

File: inc/class-img-shortcode.php [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+							'attachment' => 'Link to attachment file',
+							'file'       => 'Link to file',
+							'custom'     => 'Custom link',
+						),
+					),
+
+				),
+			);
+
+		$shortcode_ui_args = apply_filters( 'image_shortcake_ui_args', $shortcode_ui_args );
```
Comment (danielbachhuber) [inline inc/class-img-shortcode.php]: This filter name probably needs to more closely match the class name.

File: inc/class-img-shortcode.php [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+			$image_attr['src'] = $attachment[0];
+			$image_attr['width'] = $attachment[1];
+			$image_attr['height'] = $attachment[2];
+		} else {
+			$image_attr['src'] = $attr['src'];
+		}
+
+		foreach ( $image_attr as $attr_name => $attr_value ) {
+			if ( ! empty( $attr_value ) ) {
+				$image_html .= $attr_name . '="' . esc_attr( $attr_value ) . '" ';
```
Comment (danielbachhuber) [inline inc/class-img-shortcode.php]: We should probably sanitize `$attr_name` with `sanitize_key()`

**Comment by danielbachhuber** [comment]
> want to review/merge so I can start new branches for features?

Just a couple more small nits and then I think it's good to go.

**Comment by goldenapples** [comment]
Addressed changes; check again?

File: inc/class-img-shortcode.php [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+							'attachment' => 'Link to attachment file',
+							'file'       => 'Link to file',
+							'custom'     => 'Custom link',
+						),
+					),
+
+				),
+			);
+
+		$shortcode_ui_args = apply_filters( 'image_shortcode_ui_args', $shortcode_ui_args );
```
Comment (danielbachhuber) [inline inc/class-img-shortcode.php]: Given the class name is `Img_Shortcode`, can we standardize on `img_shortcode` or rename the class?

File: inc/class-img-shortcode.php (hunk 8 above)
Comment (goldenapples) [inline inc/class-img-shortcode.php]: Yup. b50ac94

File: inc/class-img-shortcode.php (hunk 9 above)
Comment (goldenapples) [inline inc/class-img-shortcode.php]: Yup, fixed in 4ae03f5.

File: inc/class-img-shortcode.php [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+	 */
+	public static function get_shortcode_ui_args() {
+
+		$shortcode_ui_args = array(
+
+				'label' => 'Image',
+
+				'listItemImage' => 'dashicons-format-image',
+
+				'post_type'     => get_post_types(),
```
Comment (danielbachhuber) [inline inc/class-img-shortcode.php]: Oh. This argument isn't required. I think we can just leave it out. Sorry I wasn't clear about that.

File: inc/class-img-shortcode.php [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+			$image_attr['src'] = $attachment[0];
+			$image_attr['width'] = $attachment[1];
+			$image_attr['height'] = $attachment[2];
+		} else {
+			$image_attr['src'] = $attr['src'];
+		}
+
+		foreach ( $image_attr as $attr_name => $attr_value ) {
+			if ( ! empty( $attr_value ) ) {
+				$image_html .= sanitize_key( $attr_name ) . '="' . esc_attr( $attr_value ) . '" ';
```
Comment (danielbachhuber) [inline inc/class-img-shortcode.php]: Oh, I missed this before, but `src` needs to be escaped with `esc_url()`

**Comment by goldenapples** [comment]
Updated, addressed more changes.

**Pull request merged**